core/amber/src/main/scala/edu/uci/ics/amber/engine/architecture/worker/ChannelMarkerManager.scala

```diff
@@ -15,54 +16,60 @@ import edu.uci.ics.amber.engine.common.virtualidentity.{
 
 import scala.collection.mutable
 
+object ChannelMarkerManager {
+  final case class MarkerContext(marker: ChannelMarkerPayload, fromChannel: ChannelIdentity)
+}
+
 class ChannelMarkerManager(val actorId: ActorVirtualIdentity, inputGateway: InputGateway)
     extends AmberLogging {
 
   private val markerReceived =
     new mutable.HashMap[ChannelMarkerIdentity, Set[ChannelIdentity]]().withDefaultValue(Set())
 
+  private var markerContext: MarkerContext = _
+
+  val checkpoints = new mutable.HashMap[ChannelMarkerIdentity, CheckpointState]()
+
+  def setContext(marker: ChannelMarkerPayload, from: ChannelIdentity): Unit = {
+    markerContext = MarkerContext(marker, from)
+  }
+
+  def getContext: MarkerContext = markerContext
```

Comment:
I personally don't like this design, which assumes each accessor of the manager uses the same context. this may become fragile. As you mentioned, handlers may not need to access additional information, we can consider remove this?

Reply:
removed the entire context.